### DISCUSSION TITLE
Is pyright too picky about empty containers?

### DISCUSSION THREAD

**Discussion by dimaqq** [microsoft/pyright#8815] [Q&A] [answered]
```py
a = lambda: {42}
b = lambda: set() # E: Return type of lambda, "set[Unknown]", is partially unknown
```

Same for literal empty dict `{}`, frozenset, etc.

I can work around that using `lambda: typing.cast(typing.Set[str], set())` but that doesn't seem too elegant.

What am I doing wrong?

**Comment by erictraut** [comment, marked as answer]
It looks like you have enabled the `reportUnknownLambda` type checking rules, which is off be default. Or more likely you've enabled "strict" mode. In this case, pyright will report when a lambda expression is evaluated with a partially-unknown type. 

If you don't want pyright to report errors of this sort, I recommend not using strict mode.

If you want to use strict mode, you will need to provide additional information to the type checker in cases like this. There are a couple of options here:

1. When calling the `set` constructor, provide explicit specialization such as `set[int]()`.
2. Provide a type annotation for the variable you are assigning the lambda to. Pyright will then use [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) to infer the intended type of the `set`. 

```python
b: Callable[[], set[int]] = lambda: set()
```
